conductor_api/src/keystore.rs

```diff
@@ -0,0 +1,763 @@
+    #[allow(dead_code)]
+    pub fn add_random_seed(&mut self, dst_id_str: &str, size: usize) -> HcResult<()> {
+        let dst_id = self.check_dst_identifier(dst_id_str)?;
+        let seed_buf = generate_random_buf(size);
+        let secret = Arc::new(Mutex::new(Secret::Seed(seed_buf)));
+        self.cache.insert(dst_id.clone(), secret);
+        self.encrypt(&dst_id)?;
+        Ok(())
+    }
+
+    fn check_dst_identifier(&self, dst_id_str: &str) -> HcResult<String> {
+        let dst_id = dst_id_str.to_string();
+        if self.secrets.contains_key(&dst_id) {
+            return Err(HolochainError::ErrorGeneric(
+                "identifier already exists".to_string(),
+            ));
+        }
+        Ok(dst_id)
+    }
+
+    /// gets a secret from the keystore
+    #[allow(dead_code)]
+    pub fn get(&mut self, src_id_str: &str) -> HcResult<Arc<Mutex<Secret>>> {
+        let src_id = src_id_str.to_string();
+        if !self.secrets.contains_key(&src_id) {
+            return Err(HolochainError::ErrorGeneric(
+                "unknown source identifier".to_string(),
+            ));
+        }
+
+        if !self.cache.contains_key(&src_id) {
+            self.decrypt(&src_id)?;
+        }
+
+        Ok(self.cache.get(&src_id).unwrap().clone()) // unwrap ok because we made sure src exists
+    }
+
+    fn check_identifiers(
+        &mut self,
+        src_id_str: &str,
+        dst_id_str: &str,
+    ) -> HcResult<(Arc<Mutex<Secret>>, String)> {
+        let dst_id = self.check_dst_identifier(dst_id_str)?;
+        let src_secret = self.get(src_id_str)?;
+        Ok((src_secret, dst_id))
+    }
+
+    /// adds a derived seed into the keystore
+    #[allow(dead_code)]
+    pub fn add_seed_from_seed(
+        &mut self,
+        src_id_str: &str,
+        dst_id_str: &str,
+        context: &SeedContext,
+        index: u64,
+    ) -> HcResult<()> {
+        let (src_secret, dst_id) = self.check_identifiers(src_id_str, dst_id_str)?;
+        let secret = {
+            let mut src_secret = src_secret.lock().unwrap();
+            match *src_secret {
+                Secret::Seed(ref mut src) => {
+                    let seed = generate_derived_seed_buf(src, context, index, SEED_SIZE)?;
+                    Arc::new(Mutex::new(Secret::Seed(seed)))
+                }
+                _ => {
+                    return Err(HolochainError::ErrorGeneric(
+                        "source secret is not a root seed".to_string(),
+                    ));
+                }
+            }
+        };
+        self.cache.insert(dst_id.clone(), secret);
+        self.encrypt(&dst_id)?;
+
+        Ok(())
+    }
+
+    /// adds a keypair into the keystore based on a seed already in the keystore
+    /// returns the public key
+    #[allow(dead_code)]
+    fn add_key_from_seed(
+        &mut self,
+        src_id_str: &str,
+        dst_id_str: &str,
+        context: &SeedContext,
+        index: u64,
```

Comment:
I still feel like we need to delete `context` and `index` from this function to remove duplication.

Implementors can make two calls:

`add_seed_from_seed("device_pin", "app_seed_42", b"HCAPPLIC", 42);`

then

`add_key_from_seed("app_seed_42", "app_sign_42", KeyType::Signing);`

Reply:
Hmmm. So are you saying that I've added an extra derivation call to `kdf::derive` (via the call to `generate_derived_seed_buf` at line 308)  which is unnecessary, and that all we need to do is used the key returned from the src seed?  I guess it would be a very easy fix.

Reply:
Ok.  Removed the extra derive.